sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java

```diff
@@ -0,0 +1,177 @@
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.io.range.ByteKey;
+import org.apache.beam.sdk.io.range.ByteKeyRange;
+import org.apache.beam.sdk.transforms.DoFn;
+
+/**
+ * A {@link RestrictionTracker} for claiming {@link ByteKey}s in a {@link ByteKeyRange} in a
+ * monotonically increasing fashion. The range is a semi-open bounded interval [startKey, endKey)
+ * where the limits are both represented by ByteKey.EMPTY.
+ */
+public class ByteKeyRangeTracker extends RestrictionTracker<ByteKeyRange, ByteKey> {
+  private ByteKeyRange range;
+  @Nullable private ByteKey lastClaimedKey = null;
+  @Nullable private ByteKey lastAttemptedKey = null;
+
+  private ByteKeyRangeTracker(ByteKeyRange range) {
+    this.range = checkNotNull(range);
+  }
+
+  /**
+   * Instantiates a new {@link ByteKeyRangeTracker} with the specified range. The keys in the range
+   * are left padded to be the same length in bytes.
+   */
+  public static ByteKeyRangeTracker of(ByteKeyRange range) {
+    return new ByteKeyRangeTracker(ByteKeyRange.of(range.getStartKey(), range.getEndKey()));
+  }
+
+  @Override
+  public synchronized ByteKeyRange currentRestriction() {
+    return range;
+  }
+
+  @Override
+  public synchronized ByteKeyRange checkpoint() {
+    checkState(lastClaimedKey != null, "Can't checkpoint before any successful claim");
+    final ByteKey nextKey = next(lastClaimedKey);
+    // hack to force range to be bigger than the range upper bundle because ByteKeyRange *start*
+    // must always be less than *end*.
+    // we don't calc next if end key is empty because it means it is the maximum upper bound []
+    ByteKey rangeEndKey =
+        nextKey.equals(range.getEndKey()) && !range.getEndKey().isEmpty()
+            ? next(range.getEndKey())
+            : range.getEndKey();
+    //    Assert that nextKey <= range.GetEndKey()
+    ByteKeyRange res = ByteKeyRange.of(nextKey, rangeEndKey);
+    this.range = ByteKeyRange.of(range.getStartKey(), nextKey);
+    return res;
+  }
+
+  /**
+   * Attempts to claim the given key.
+   *
+   * <p>Must be larger than the last successfully claimed key.
+   *
+   * @return {@code true} if the key was successfully claimed, {@code false} if it is outside the
+   *     current {@link ByteKeyRange} of this tracker (in that case this operation is a no-op).
+   */
+  @Override
+  protected synchronized boolean tryClaimImpl(ByteKey key) {
+    checkArgument(
+        lastAttemptedKey == null || key.compareTo(lastAttemptedKey) > 0,
+        "Trying to claim key %s while last attempted was %s",
+        key,
+        lastAttemptedKey);
+    checkArgument(
+        key.compareTo(range.getStartKey()) > -1,
```

Comment:
<!--new_thread; commit:d28b9222a4c730ab557091334c555aa8fe89bc67; resolved:0-->
Optimization: I think comparing to startKey is only necessary if lastAttemptedKey == null